src/tools/idlc/src/tt_create.c

```diff
@@ -0,0 +1,517 @@
+/*
+ * Copyright(c) 2006 to 2019 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+#include "dds/ddsrt/retcode.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/string.h"
+#include "dds/ddsts/typetree.h"
+#include "tt_create.h"
+
+
+typedef struct array_size array_size_t;
+struct array_size {
+  unsigned long long size;
+  array_size_t *next;
+};
+
+struct dds_context {
+  ddsts_type_t *root_type;
+  ddsts_type_t *cur_type;
+  ddsts_type_t *type_for_declarator;
+  array_size_t *array_sizes;
+  dds_retcode_t retcode;
+  ddsts_identifier_t dangling_identifier;
+};
+
+extern bool dds_new_base_type(dds_context_t *context, ddsts_flags_t flags, ddsts_type_t **result)
```

Comment:
I did consider using dds_retcode_t. However, the generated parser, from which these functions are called, cannot return a dds_retcode_t. It returns one of the integer values: 0 (succes), 1 (parser error), or 2 (out-of-memory). So, the only way to get the return code outside the parser is by storing it in the context. The functions are only intended to be called from idl.y. I thought it would result in cleaner code to store the return code in the context inside the function, than having to do this in the idl.y file.